### DISCUSSION TITLE
isearch + magit-status = frozen Emacs

### DISCUSSION THREAD

**Discussion by rudolf-adamkovic** [magit/magit#4608] [Support]
I would like to quickly navigate `magit-status` with `isearch`, but almost every time I try, my Emacs freezes to death.  I figured that I can make `isearch` avoid searching through invisible content with

```emacs-lisp
(add-hook 'magit-status-mode-hook
          (lambda () (setq-local search-invisible nil)))
```

This works in other modes, but it seems to "do nothing" in `magit-status`.

Any ideas?

P.S. The freezes happen even with a relatively small amount of data, e.g. 20 megabytes in the buffer, according to the `size-indication-mode`.  The computer has 16 gigabytes of main memory and a modern quad-core processor.

[Other operations in the `magit-status` buffer have "syrupy" performance, but they at least do not freeze the computer.  Hence, I report this issue as a bug, for I cannot search at all.]

**Comment by tarsius** [comment]
> Hence, I report this issue as a bug, for I cannot search at all.

Because I don't think there is anything that can be done about it on our end, I am converting it to a discussion.

> Any ideas?

All I've got is "try to avoid this situation".

20mb is not nothing.

If this involves many untracked files, then setup good ignore rules (prefer ignoring directories over file extensions).

> I figured that I can make isearch avoid searching through invisible content 

Emacs still has to find the end of the invisible portion when it hits the beginning. This may involve very long lines from, e.g., uncommitted minified js. Emacs doesn't deal well with long lines. `solong-mode` helps with that by turning off features, but that would break Magit. Try to commit the generated things immediately. If yet to be handwritten changes belong in the same commit, you can later amend.

**Reply by rudolf-adamkovic** [reply]
Ignoring performance for a moment, could Magit respect `C-s M-s i` (search just the visible content), like other modes do?

**Reply by tarsius** [reply]
I don't use isearch myself, so I had to use `M-x isearch-forward`. Then I pressed `M-s i` but that is unbound. I looked at `isearch-mode-map` and there is no binding for `M-s i`. What is that bound to for you?

**Reply by tarsius** [reply]
I searched a bit and found `isearch-invisible`. I set that to nil and then isearch didn't look into text hidden by an overlay anymore. So yes, magit can do that like other modes.

**Reply by rudolf-adamkovic** [reply]
> I don't use isearch myself, so I had to use `M-x isearch-forward`. Then I pressed `M-s i` but that is unbound. I looked at `isearch-mode-map` and there is no binding for `M-s i`. What is that bound to for you?

`isearch-toggle-invisible`

See: https://www.gnu.org/software/emacs/manual/html_node/emacs/Special-Isearch.html

**Reply by rudolf-adamkovic** [reply]
> I searched a bit and found `isearch-invisible`. I set that to nil and then isearch didn't look into text hidden by an overlay anymore. So yes, magit can do that like other modes.

I use `isearch` to navigate around a lot (but not in Magit, to avoid freezes).  When I set `isearch-invisible` from `open` to `nil`, `isearch` still "opens" (unfolds) the files in `magit-status`.

The documentation says:

```emacs-lisp
(defcustom search-invisible 'open
  "If t incremental search/query-replace can match hidden text.
A nil value means don't match invisible text.
When the value is `open', if the text matched is made invisible by
an overlay having a non-nil `invisible' property, and that overlay
has a non-nil property `isearch-open-invisible', then incremental
search will show the hidden text.  (This applies when using `outline.el'
and `hideshow.el'.)

To temporarily change the value for an active incremental search,
use \\<isearch-mode-map>\\[isearch-toggle-invisible].

See also the related option `isearch-hide-immediately'.

See also `reveal-mode' if you want overlays to automatically be opened
whenever point is in one of them."
  :type '(choice (const :tag "Match hidden text" t)
		 (const :tag "Open overlays" open)
		 (const :tag "Don't match hidden text" nil)))
```

P.S. What do you use instead of `isearch`?

**Reply by tarsius** [reply]
Note that there are two variables `isearch-invisible` and `search-invisible`.  You have to set the former. For me that works when using isearch.

**Reply by rudolf-adamkovic** [reply]
> Note that there are two variables `isearch-invisible` and `search-invisible`. You have to set the former. For me that works when using isearch.

All right, something does not add up here.  I have just launched Emacs with `emacs -Q`, added MELPA, installed `magit`, and then tried it.  I opened `magit-status`, folded everything, and then searched for a word.  I used `C-s M-s i` to toggle `isearch-invisible` to `nil` (and verified it with `C-h v` while in `isearch`) right before I entered the search term.  Magit froze for 20 seconds and then unfolded the various sections down to the diff with the word I searched for.  I run vanilla GNU Emacs 29, pulled and compiled 5 days ago from the `master` branch.

**Reply by tarsius** [reply]
Try with this modified function:

```elisp
(defun magit-section--open-temporarily (beg end)
  (or (not (isearch-range-invisible beg end))
      (and (eq search-invisible t)
           (progn
             (save-excursion
               (goto-char beg)
               (let ((section (magit-current-section)))
                 (while section
                   (let ((content (oref section content)))
                     (if (and (magit-section-invisible-p section)
                              (<= (or content (oref section start))
                                  beg
                                  (oref section end)))
                         (progn
                           (when content
                             (magit-section-show section)
                             (push section magit-section--opened-sections))
                           (setq section (oref section parent)))
                       (setq section nil))))))
             (not (isearch-range-invisible beg end))))))
```

**Reply by rudolf-adamkovic** [reply]
It did not work, but after a bit of debugging, I found that changing `(eq search-invisible t)` to `(eq isearch-invisible t)` makes it work.  I can now `C-s M-s i` and Emacs searches the Magit buffer super-fast, without opening the sections!  Will this become a part of Magit proper or should I set some time aside to hack it into my configuration somehow?

**Reply by tarsius** [reply]
It works with the modified function + the changed variable, or with just the changed variable?

I'll add it to magit.

**Reply by rudolf-adamkovic** [reply]
> It works with the modified function + the changed variable, or with just the changed variable?

The modified function does _not_ work "as is", but everything starts working when `(eq search-invisible t)` in the modified function becomes `(eq isearch-invisible t)`.

> I'll add it to magit.

Thank you so much!  For me, this will represent the most impactful change in Magit since I started using it long ago.

_edit: clarity_

**Reply by tarsius** [reply]
Turns out this is way more complex than expected. I still intend to fix this properly.

Meanwhile something like this should work:

```lisp
(advice-add 'magit-section--open-temporarily :override
            magit-section--open-temporarily)
```
